File: tools/gui/src/domain/schemas/changes.ts
import { Schema } from 'effect'

/**
 * One local change: an edit to an artifact this repository does not own, held here until it is
 * submitted for review upstream.
 *
 * `artifact_id` is the change's own — an artifact may carry more than one, so it is what discarding
 * names. `target_id` is the promoted artifact, which is what a link goes to and what a reader knows.
 * The local reference standing for it is machinery and is deliberately not published.
 */
/**
 * One field of a stale change: what it asks for, and what the artifact says now.
 *
 * Both null where the value has no single line to show — a properties table, an attribute-type map.
 * Naming the field is still worth saying; a rendering invented for the list would be a second,
 * worse spelling of what the artifact view already draws.
 */
export const FieldDivergenceSchema = Schema.Struct({
  field: Schema.String,
  proposed: Schema.NullOr(Schema.String),
  current: Schema.NullOr(Schema.String),
})
export type FieldDivergence = typeof FieldDivergenceSchema.Type

export const ChangeSummarySchema = Schema.Struct({
  artifact_id: Schema.String,
  target_id: Schema.String,
  target_name: Schema.String,
  kind: Schema.Literal('entity', 'document', 'diagram'),
  changed_fields: Schema.Array(Schema.String),
  state: Schema.Literal('draft', 'submitted'),
  // Computed on read: `stale` means the enterprise artifact has moved since the change was written.
  condition: Schema.Literal('current', 'stale', 'conflicting'),
  /** What the change asks for beside what the artifact says now. Empty unless it is stale. */
  divergence: Schema.Array(FieldDivergenceSchema),
})
export type ChangeSummary = typeof ChangeSummarySchema.Type

export const ChangeListSchema = Schema.Struct({
  changes: Schema.Array(ChangeSummarySchema),
  total: Schema.Number,
})
export type ChangeList = typeof ChangeListSchema.Type

export const ChangeDiscardedSchema = Schema.Struct({
  artifact_id: Schema.String,
  discarded: Schema.Boolean,
  state: Schema.Literal('abandoned'),
})
export type ChangeDiscarded = typeof ChangeDiscardedSchema.Type

/**
 * Where one change stood when a rebase rehearsed it, and what was done about it.
 *
 * `reason` is what a reader acts on — for a conflict it is the verifier's own refusal, because a
 * second wording of a refusal is a second vocabulary for the same thing.
 */
export const RebasedChangeSchema = Schema.Struct({
  artifact_id: Schema.String,
  target_id: Schema.String,
  outcome: Schema.Literal('clean', 'superseded', 'conflicting'),
  reason: Schema.String,
  restamped: Schema.Boolean,
})
export type RebasedChange = typeof RebasedChangeSchema.Type

export const ChangeRebasedSchema = Schema.Struct({
  changes: Schema.Array(RebasedChangeSchema),
  /**
   * The replacement review branch a rebase of a submitted set was published on. Null for a draft,
   * which has no published branch to replace — the branch is the unit of review, so a set already
   * under review gets a new one rather than having the one a reviewer is reading rewritten.
   */
  republished_branch: Schema.NullOr(Schema.String),
  summary: Schema.String,
})
export type ChangeRebased = typeof ChangeRebasedSchema.Type

File: src/infrastructure/rest/contracts/changes.py
"""Response contracts for the local changes this repository is holding.

A change row names the promoted artifact: its id and what it is called. Not the local reference
standing for it — that is machinery, excluded from every list and every search, and a row publishing
one invited a client to link it. The artifact a reader knows is the promoted one, which is what
search returns and what they edited.
"""

from __future__ import annotations

from typing import Literal

from src.infrastructure.rest.contracts.wire_shape import Closed


class FieldDivergence(Closed):
    """One field of a stale change: what it asks for, and what the artifact says now.

    Both null where the value has no single line to show — a properties table, an attribute-type
    map. Naming the field as diverging is still worth saying; rendering a structured value here
    would be a second, worse spelling of what the artifact view already draws.
    """

    field: str
    proposed: str | None
    current: str | None


class ChangeSummary(Closed):
    """One local change: what it changes, what it says, and where it stands."""

    #: The change's own id. What a caller names to discard it — two changes against one artifact are
    #: ordinary, so the artifact is not enough to identify one.
    artifact_id: str
    target_id: str
    target_name: str
    kind: Literal["entity", "document", "diagram"]
    changed_fields: list[str]
    state: Literal["draft", "submitted"]
    #: `stale` means the enterprise artifact has moved since this was written. Computed on read, so
    #: it is never a claim the file makes about the world.
    condition: Literal["current", "stale", "conflicting"]
    #: What the change asks for beside what the artifact says now. Empty unless the change is
    #: stale: on a current one the two agree, and a value shown beside itself is noise.
    divergence: list[FieldDivergence]


class ChangeListResponse(Closed):
    """Every live change, ordered by the artifact each one changes."""

    changes: list[ChangeSummary]
    total: int


class ChangeDiscardedResponse(Closed):
    """What a discard did.

    The record is kept and moved to a terminal state rather than deleted: a change that was submitted
    has been seen by someone, and its disappearance would be indistinguishable from it never having
    existed. `discarded` is false when the change was already terminal, which is not an error.
    """

    artifact_id: str
    discarded: bool
    state: Literal["abandoned"]


class RebasedChange(Closed):
    """Where one change stood when the rebase rehearsed it, and what was done about it.

    `outcome` is `change_rebase`'s own vocabulary, and `reason` is what an author reads — for a
    conflict it is the verifier's refusal verbatim, because a second wording of a refusal is a
    second vocabulary for the same thing.
    """

    artifact_id: str
    target_id: str
    outcome: Literal["clean", "superseded", "conflicting"]
    reason: str
    #: Whether the change now records the revision it was just proven against. True only for a clean
    #: outcome, and only where the artifact's current revision could be read.
    restamped: bool


class ChangeRebasedResponse(Closed):
    """What the rehearsal concluded. Nothing is written for a superseded or conflicting change."""

    changes: list[RebasedChange]
    summary: str
